File: java/tea-framework/src/main/java/com/tea/framework/parser/TeaFile.java
package com.tea.framework.parser;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class TeaFile {
    public String getPath() {
        return path;
    }

    public List<Token> getTokens() {
        return tokens;
    }

    public String getUrl() {
        return url;
    }

    public String getTestCase() {
        return testCase;
    }

    public BrowserActionLine getLines() {
        return lines;
    }

    public List<ActionLine> getActionLines() {
        return lines.actionLines;
    }

    public void setPath(String path) {
        this.path = path;
    }

    private String path;
    private List<Token> tokens;
    private String url;
    private String testCase;
    private BrowserActionLine lines;

    public TeaFile(String path) throws FileNotFoundException {
        this.path = path;
        tokens = new ArrayList<Token>();
        ReadFile();
        ParseTokens();
    }

    private void ReadFile() throws FileNotFoundException {
        // Read the file and display it line by line.  
        Scanner file = new Scanner(new File(path));
        int count = 0;
        while (file.hasNextLine()) {
            count++;
            String line = file.nextLine();
            if (line.isEmpty()) continue;

            Lexer lexer = new Lexer(new SourceScanner(line));

            Token token = null;
            try {
                token = lexer.readNext();
            } catch (IllegalArgumentException e) {
                e.printStackTrace();
                throw new IllegalStateException("Failed to read line " + count + ".\n");
            }

            while (token != null && token.getType() != Token.TokenType.TOKEN_EOE) {
                tokens.add(token);
                token = lexer.readNext();
            }
        }

        file.close();
    }

    private void ParseTokens() {

        for (Token t : getTokens()) {
            if (t.getValue() != null && t.getValue().length() > 0) {
                // assign Url
                if (t.getType() == Token.TokenType.TOKEN_URL) {
                    url = t.getValue();
                    continue;
                }

                // assign testCase
                if (t.getType() == Token.TokenType.TOKEN_TC_DESC) {
                    testCase = t.getValue();
                    continue;
                }
            }
        }
        lines = new BrowserActionLine(tokens);
    }
}

File: dotnet-core/Tea/Parsing/TeaParser/TeaFile.cs
/*
    A teafile is a text file ending in .tea.  Teafiles will contain all the information required to run selenium test automation and actions. The Teafile class stores all the relevant information for the selenium architecture to run.

    A teafile consists of a url, test case keyword, and a series of actions- e.g.,

    /// LoginWithAFailingUsername.tea
    http://www.reddit.com

    Test Case: Trying to login to reddit with a random username/password will fail.
        click   xpath   //*[text() = 'Log In'] # Click login button
        click   xpath   //*[@id='loginUsername']
        type    id      random
        click   id      loginPassword
        type    id      password
        click   xpath   //fieldset[5]/button
*/

using System.Collections.Generic;

namespace TeaParser
{

    public class TeaFile
    {
        private string Path {get; set;}
        public List<Token> Tokens {get; private set;}
        public string URL { get; private set; }
        public string TestCase { get; private set; }
        public BrowserActionLine Lines {get; private set; }
        public List<ActionLine> ActionLines {
             get
             {
                  return Lines.ActionLines;
             }
            private set {} }
        public TeaFile(string path)
        {
            Path = path;
            Tokens = new List<Token>();
            ReadFile();
            ParseTokens();
        }

        private void ReadFile()
        {
            // Read the file and display it line by line.  
            System.IO.StreamReader file = new System.IO.StreamReader(Path);  
            string line;
            while((line = file.ReadLine()) != null)
            {
                var lexer = new TeaParser.Lexer(new TeaParser.SourceScanner(line));
                
                Token token = lexer.ReadNext();
                while ( token != null && token.Type != Token.TokenType.TOKEN_EOE)
                {
                    Tokens.Add(token);
                    token = lexer.ReadNext();
                }
                
            }

            file.Close();
        }

        private void ParseTokens()
        {
            
            foreach (var t in Tokens)
            {
                if ( t.Value != null && t.Value.Length > 0)
                {
                    // assign Url
                    if ( t.Type == Token.TokenType.TOKEN_URL ) { URL = t.Value; continue; }

                    // assign testCase
                    if ( t.Type == Token.TokenType.TOKEN_TC_DESC ) { TestCase = t.Value; continue; }
                }
            }
            Lines = new BrowserActionLine(Tokens);
        }
        
    }
}